[CodeCompletion] Don’t report a call pattern as convertible if its result type doesn’t match

ahoppen opened the pull request:
To compute the expected type of a call pattern (which is the return type of the function if that call pattern is being used), we called `getTypeForCompletion` for the entire call, in the same way that we do for the code completion token. However, this pattern does not generally work. For the code completion token it worked because the code completion expression doesn’t have an inherent type and it inherits the type solely from its context. Calls, however, have an inherent return type and that type gets assigned as the `typeForCompletion`.

Implement targeted checks for the two most common cases where an expected type exists: If the call that we suggest call patterns for is itself an argument to another function or if it is used in a place that has a contextual type in the constraint system (eg. a variable binding or a `return` statement). This means that we no longer return `Convertible` for call patterns in some more complex scenarios. But given that this information was computed based on incorrect results and that in those cases all call patterns had a `Convertible` type relation, I think that’s acceptable. Fixing this would require recording more information in the constraints system, which is out-of-scope for now.

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -5947,8 +5930,27 @@ bool ConstraintSystem::repairFailures(
+        // other (explicit) argument's so source range containment alone isn't
+        // sufficient.
+        bool isSynthesizedArg = arg->isImplicit() && isa<DeclRefExpr>(arg);
+        if (!isSynthesizedArg && isForCodeCompletion() &&
```
nit: `isForCodeCompletion()` is already checked at the outer `if`

rintaro reviewed (approved):
Looks reasonable

ahoppen commented:
@swift-ci Please smoke test

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -5947,8 +5930,26 @@ bool ConstraintSystem::repairFailures(
+        bool isSynthesizedArg = arg->isImplicit() && isa<DeclRefExpr>(arg);
+        if (!isSynthesizedArg && containsIDEInspectionTarget(arg) &&
+            !lhs->isVoid() && !lhs->isUninhabited())
+          fixBehavior = FixBehavior::Suppress;
```
I replied about this in the other PR, I think we need a better way here.

ahoppen committed:
[CodeCompletion] Don’t report a call pattern as convertible if its result type doesn’t match

To compute the expected type of a call pattern (which is the return type of the function if that call pattern is being used), we called `getTypeForCompletion` for the entire call, in the same way that we do for the code completion token. However, this pattern does not generally work. For the code completion token it worked because the code completion expression doesn’t have an inherent type and it inherits the type solely from its context. Calls, however, have an inherent return type and that type gets assigned as the `typeForCompletion`.

Implement targeted checks for the two most common cases where an expected type exists: If the call that we suggest call patterns for is itself an argument to another function or if it is used in a place that has a contextual type in the constraint system (eg. a variable binding or a `return` statement). This means that we no longer return `Convertible` for call patterns in some more complex scenarios. But given that this information was computed based on incorrect results and that in those cases all call patterns had a `Convertible` type relation, I think that’s acceptable. Fixing this would require recording more information in the constraints system, which is out-of-scope for now.

ahoppen commented:
@swift-ci Please smoke test

rintaro reviewed (approved):
> we no longer return Convertible for call patterns in some more complex scenarios.

let's improve this in follow-ups
